azext_iot/monitor/property.py

```diff
@@ -7,91 +7,207 @@
-    polling_interval_seconds=DEVICETWIN_POLLING_INTERVAL_SEC,
-):
-    prev_twin = None
+        changes = {
+            key: self._changed_props(prop.props[key], prop.metadata[key], key,)
+            for key, val in prop.metadata.items()
+            if self._is_relevant(key, val)
+        }
 
-    device_twin_provider = CentralDeviceTwinProvider(
-        cmd=cmd, app_id=app_id, device_id=device_id
-    )
+        return changes
 
-    while True:
+    def _is_relevant(self, key, val):
+        if key in {"$lastUpdated", "$lastUpdatedVersion"}:
+            return False
 
-        raw_twin = device_twin_provider.get_device_twin(
-            central_dns_suffix=central_dns_suffix
+        updated_within = datetime.datetime.now() - datetime.timedelta(
+            seconds=DEVICETWIN_MONITOR_TIME_SEC
         )
 
-        twin = DeviceTwin(raw_twin)
-        if prev_twin:
-            change_d = compare_properties(
-                prev_twin.desired_property, twin.desired_property
+        last_updated = isodate.parse_datetime(val["$lastUpdated"])
+        return last_updated.timestamp() >= updated_within.timestamp()
+
+    def _changed_props(self, prop, metadata, property_name):
+        # not an interface - whole thing is change log
+        if not self._is_interface(property_name):
+            return prop
+        # iterate over property in the interface
+        # if the property is not an exact match for what is present in the previous set of properties
+        # track it as a change
+        diff = {
+            key: prop[key]
+            for key, val in metadata.items()
+            if self._is_relevant(key, val)
+        }
+        return diff
+
+    def _validate_payload(self, changes, minimum_severity):
+        for value in changes:
+            issues = self._validate_payload_against_interfaces(
+                changes[value], value, minimum_severity
+            )
+            for issue in issues:
+                issue.log()
+
+    def _validate_payload_against_interfaces(
+        self, payload: dict, name, minimum_severity
+    ):
+        name_miss = []
+        issues_handler = IssueHandler()
+        interface_name = name.replace(CENTRAL_PNP_INTERFACE_PREFIX, "")
+        if self._is_interface(interface_name):
+            # if the payload is an interface then iterate thru the properties under the interface
+            for property_name in payload.keys():
+                schema = self._template.get_schema(
+                    name=property_name, interface_name=interface_name
+                )
+                if not schema:
+                    name_miss.append(property_name)
+        else:
+            # if the payload is a property then process the payload as a single unit.
+            schema = self._template.get_schema(name=name, interface_name="")
+
+            if not schema:
+                name_miss.append(name)
+
+            interfaces_with_specified_property = self._validate_duplicate_properties(
```

Comment:
maintenance - this isn't validating duplicate properties, its finding a list of interfaces that contain a schema matching "name"

Requested action: rename the function
Suggested action: move function to Template class